src/gui/mzroll/librarymanager.cpp

```diff
@@ -0,0 +1,333 @@
+    LibraryRecord databaseRecord(filepath,
+                                 QString::fromStdString(dbName),
+                                 static_cast<int>(dbCompounds.size()));
+    return databaseRecord;
+}
+
+LibraryWorker* LibraryManager::_newWorker()
+{
+    LibraryWorker *worker = new LibraryWorker;
+    worker->moveToThread(&_workerThread);
+    connect(&_workerThread, &QThread::finished, worker, &QObject::deleteLater);
+    _workerThread.start();
+    return worker;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+LibraryRecord::LibraryRecord(const QJsonObject& json)
+{
+    absolutePath = "";
+    databaseName = "";
+    numberOfCompounds = 0;
+
+    if (json.contains("absolutePath") && json["absolutePath"].isString())
+        absolutePath = json["absolutePath"].toString();
+
+    if (json.contains("databaseName") && json["databaseName"].isString())
+        databaseName = json["databaseName"].toString();
+
+    if (json.contains("numberOfCompounds") && json["numberOfCompounds"].isDouble())
+        numberOfCompounds = json["numberOfCompounds"].toInt();
+}
+
+QJsonObject LibraryRecord::toJson() const
+{
+    QJsonObject json;
+    json["absolutePath"] = absolutePath;
+    json["databaseName"] = databaseName;
+    json["numberOfCompounds"] = numberOfCompounds;
+    return json;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+void LibraryWorker::saveRecord(const LibraryRecord& record)
+{
+    bool recordAlreadyExists = false;
+    auto entries = _loadEntries();
+    for (size_t i = 0; i < entries.size(); ++i) {
+        QJsonObject entry = entries[i].toObject();
+        LibraryRecord existingRecord(entry);
+        // record having the exact same file path should have the same stats
+        if (existingRecord.absolutePath == record.absolutePath) {
+            existingRecord.databaseName = record.databaseName;
+            existingRecord.numberOfCompounds = record.numberOfCompounds;
+            entries.replace(i, existingRecord.toJson());
+            recordAlreadyExists = true;
+        }
+    }
+
+    if (!recordAlreadyExists)
+        entries.append(record.toJson());
+
+    emit saveFinished(_saveEntries(entries));
+}
+
+void LibraryWorker::deleteRecord(const LibraryRecord& record)
+{
+    QList<size_t> recordsIndicesToBeDeleted;
+    auto entries = _loadEntries();
+    for (size_t i = 0; i < entries.size(); ++i) {
+        QJsonObject entry = entries[i].toObject();
+        LibraryRecord existingRecord(entry);
+        // any record having the exact same file path will be deleted
+        if (existingRecord.absolutePath == record.absolutePath)
+            recordsIndicesToBeDeleted.append(i);
+    }
+
+    if (recordsIndicesToBeDeleted.isEmpty()) {
+        emit deleteFinished(false);
+        return;
+    }
+
+    for (auto index : recordsIndicesToBeDeleted)
+        entries.removeAt(index);
+
+    emit deleteFinished(_saveEntries(entries));
```

Comment:
Same question as above for `deleteFinished` signal

Reply:
Same reason as the one for `saveFinished` given above. Let me know if you think they're not needed at all.